client/widmgr/widmgr.go

```diff
@@ -4,80 +4,367 @@
+// goroutine was not started.
+func (m *WIDMgr) Run() error {
+	if len(m.widSpecs) == 0 {
+		m.logger.Debug("no workload identities to retrieve or renew")
+		return nil
 	}
-	reply := structs.AllocIdentitiesResponse{}
-	if err := m.rpc.RPC("Alloc.SignIdentities", &args, &reply); err != nil {
-		return nil, err
+
+	m.logger.Debug("retrieving and renewing workload identities", "num_identities", len(m.widSpecs))
+
+	if err := m.getIdentities(); err != nil {
+		return fmt.Errorf("failed to fetch signed identities: %w", err)
 	}
 
-	if n := len(reply.Rejections); n == 1 {
-		return nil, fmt.Errorf("%d/%d signing request was rejected", n, len(req))
-	} else if n > 1 {
-		return nil, fmt.Errorf("%d/%d signing requests were rejected", n, len(req))
+	go m.renew()
+
+	return nil
+}
+
+// Get retrieves the latest signed identity or returns an error. It must be
+// called after Run and does not block.
+//
+// For retrieving tokens which might be renewed callers should use Watch
+// instead to avoid missing new tokens retrieved by Run between Get and Watch
+// calls.
+func (m *WIDMgr) Get(id cstructs.TaskIdentity) (*structs.SignedWorkloadIdentity, error) {
+	token := m.get(id)
+	if token == nil {
+		// This is an error as every identity should have a token by the time Get
+		// is called.
+		return nil, fmt.Errorf("unable to find token for task %q and identity %q", id.TaskName, id.IdentityName)
 	}
 
-	if len(reply.SignedIdentities) == 0 {
-		return nil, fmt.Errorf("empty signed identity response")
+	return token, nil
+}
+
+func (m *WIDMgr) get(id cstructs.TaskIdentity) *structs.SignedWorkloadIdentity {
+	m.lastTokenLock.Lock()
+	defer m.lastTokenLock.Unlock()
+
+	return m.lastToken[id]
+}
+
+// Watch sends new signed identities until it is closed due to shutdown. Must
+// be called after Run.
+//
+// The caller must call the returned func to stop watching.
+func (m *WIDMgr) Watch(id cstructs.TaskIdentity) (<-chan *structs.SignedWorkloadIdentity, func()) {
+	m.watchersLock.Lock()
+	defer m.watchersLock.Unlock()
+
+	// If Shutdown has been called return a closed chan
+	if m.stopCtx.Err() != nil {
+		c := make(chan *structs.SignedWorkloadIdentity)
+		close(c)
+		return c, func() {}
+	}
+
+	//TODO I think we only want 1 watcher per id, so let's just close on double watches?
+	if existing, ok := m.watchers[id]; ok {
+		close(existing)
+	}
+
+	// Buffer of 1 so sends don't block on receives
+	c := make(chan *structs.SignedWorkloadIdentity, 1)
+	m.watchers[id] = c
+
+	// Create a cancel func for watchers to deregister when they exit.
+	cancel := func() {
+		m.watchersLock.Lock()
+		defer m.watchersLock.Unlock()
+
+		delete(m.watchers, id)
+
+		//TODO should we close(c) here? seems sketchy... caller shouldn't try to
+		//recv on c after calling cancel
+	}
+
+	// Prime chan with latest token to avoid a race condition where consumers
+	// could miss a token update between Get and Watch calls.
```

Comment:
I really like how we take advantage of the lock in the widmgr to prime/pop values on the channel and eliminate a whole class of race conditions